File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    private void solve(List<String> ans, String curr, int n, int open, int close){
        if(curr.length() == 2*n){
            ans.add(curr);
            return;
        }

        if(open < n){
            curr += '(';
            solve(ans, curr, n, open+1, close);
            curr = curr.substring(0, curr.length()-1);
        }

        if(close < open){
            curr += ')';
            solve(ans, curr, n, open, close+1);
            curr = curr.substring(0, curr.length()-1);
        }
    }

    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();

        solve(ans, "", n, 0, 0);
        return ans;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
private:
    bool isValid(string curr){
        int count = 0;

        for(char ch : curr){
            if(ch == '(') count++;
            else count--;
            
            if(count < 0) return false;
        }

        return (count == 0);
    }
    void solve(string curr, int n, vector<string>& ans){
        if(curr.length() == 2*n){
            if(isValid(curr)){
                ans.push_back(curr);
            }

            return;
        }

        curr += '(';
        solve(curr, n, ans);
        curr.pop_back();

        curr += ')';
        solve(curr, n, ans);
        curr.pop_back();
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        solve("", n, ans);
        return ans;
    }
};